x/crosschain/keeper/grpc_query_cctx.go

```diff
@@ -177,3 +177,206 @@ func (k Keeper) CctxListPending(c context.Context, req *types.QueryListCctxPendi
 		TotalPending: totalPending,
 	}, nil
 }
+
+// CctxListPendingWithinRateLimit returns a list of pending cctxs that do not exceed the outbound rate limit
+// a limit for the number of cctxs to return can be specified or the default is MaxPendingCctxs
+func (k Keeper) CctxListPendingWithinRateLimit(c context.Context, req *types.QueryListCctxPendingWithRateLimitRequest) (*types.QueryListCctxPendingWithRateLimitResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	// check limit and use default MaxPendingCctxs if not specified
+	if req.Limit > MaxPendingCctxs {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("limit exceeds max limit of %d", MaxPendingCctxs))
+	}
+	limit := req.Limit
+	if limit == 0 {
+		limit = MaxPendingCctxs
+	}
+
+	// get current height and tss
+	ctx := sdk.UnwrapSDKContext(c)
+	height := ctx.BlockHeight()
+	if height <= 0 {
+		return nil, status.Error(codes.OutOfRange, "height out of range")
+	}
+	tss, found := k.zetaObserverKeeper.GetTSS(ctx)
+	if !found {
+		return nil, status.Error(codes.Internal, "tss not found")
+	}
+
+	// check rate limit flags to decide if we should apply rate limit
+	applyLimit := true
+	rateLimitFlags, found := k.GetRatelimiterFlags(ctx)
+	if !found || !rateLimitFlags.IsEnabled {
+		applyLimit = false
+	}
+
+	// calculate the rate limiter sliding window left boundary (inclusive)
+	leftWindowBoundary := height - rateLimitFlags.RateLimitWindow
+	if leftWindowBoundary < 0 {
+		leftWindowBoundary = 0
+	}
+
+	// get the conversion rates for all foreign coins
+	var gasCoinRates map[int64]*big.Float
+	var erc20CoinRates map[string]*big.Float
+	var rateLimitInZeta *big.Float
+	if applyLimit {
+		gasCoinRates, erc20CoinRates = k.GetRatelimiterRates(ctx)
+		rateLimitInZeta = new(big.Float).SetInt(rateLimitFlags.RateLimitInZeta.BigInt())
+	}
+
+	// define a few variables to be used in the below loops
+	limitExceeded := false
+	totalPending := uint64(0)
+	totalCctxValueInZeta := big.NewFloat(0)
+	cctxs := make([]*types.CrossChainTx, 0)
+
+	// the criteria to stop adding cctxs to the result
+	maxCCTXsReached := func() bool {
+		// #nosec G701 len always positive
+		return uint32(len(cctxs)) >= limit
+	}
+
+	// query pending cctxs for each supported chain
+	chains := k.zetaObserverKeeper.GetSupportedChains(ctx)
+ChainLoop:
+	for _, chain := range chains {
```

Comment:
It's one chain after another? So the first chains would take everything if the limit is exceeded?

Reply:
for now, yeah